### PR TITLE
Optimize finalizer addition

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area scalability
/kind enhancement

**What this PR does / why we need it**:

This PR optimizes how gardener adds finalizers on k-api and g-api resources:
- switch to strategic merge patch for adding finalizers to make use of the `merge` directive on `metadata.finalizers` and run without optimistic locking
- don't patch object if finalizer is already present

The goal is to minimize conflicts, especially on the start of reconciliations.
In our dev landscape, I observed the following numbers of conflicts caused by finalizer addition from the logs of gardener-controller-manager and gardenlet over 6 hours:
```
$ cat /tmp/conflicts-gardenlet | grep -i 'finalizer' | sed -E 's/^.*Operation cannot be fulfilled on ([^ ]*).*$/\1/' | sort | uniq -c
     10 backupentries.core.gardener.cloud
     53 shoots.core.gardener.cloud
$ cat /tmp/conflicts-gcm | grep -i 'finalizer' | sed -E 's/^.*Operation cannot be fulfilled on ([^ ]*).*$/\1/' | sort | uniq -c
   1432 plants.core.gardener.cloud
```

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/2822

**Special notes for your reviewer**:
/squash

Note the comment on `StrategicMergePatchAddFinalizers`:

> we can't do the same for removing finalizers, because removing the last finalizer results in the following patch: `{"metadata":{"finalizers":null}}` which is not safe to issue without optimistic locking.
> Also, `$deleteFromPrimitiveList` is not idempotent, see https://github.com/kubernetes/kubernetes/issues/105146.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/controllermanager/controller/managedseedset/managedseedset_reconciler.go [hunk 1]
```diff
@@ -75,9 +75,11 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 
 func (r *reconciler) reconcile(ctx context.Context, set *seedmanagementv1alpha1.ManagedSeedSet) (result reconcile.Result, err error) {
 	// Ensure gardener finalizer
-	r.getLogger(set).Debug("Ensuring finalizer")
-	if err := controllerutils.PatchAddFinalizers(ctx, r.gardenClient.Client(), set, gardencorev1beta1.GardenerName); err != nil {
-		return reconcile.Result{}, fmt.Errorf("could not ensure gardener finalizer: %w", err)
+	if !controllerutil.ContainsFinalizer(set, gardencorev1beta1.GardenerName) {
+		r.getLogger(set).Debug("Ensuring finalizer")
```
Comment (stoyanr) [inline pkg/controllermanager/controller/managedseedset/managedseedset_reconciler.go]: The original idea of "ensuring" instead of "adding" was: add the finalizer if it doesn't exist and do nothing otherwise. As such, the log message above should either be moved to the line before the check, or changed to `Adding finalizer`.

File: pkg/gardenlet/controller/managedseed/managedseed_reconciler.go [hunk 2]
```diff
@@ -70,9 +70,11 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 
 func (r *reconciler) reconcile(ctx context.Context, ms *seedmanagementv1alpha1.ManagedSeed) (result reconcile.Result, err error) {
 	// Ensure gardener finalizer
-	r.getLogger(ms).Debug("Ensuring finalizer")
-	if err := controllerutils.PatchAddFinalizers(ctx, r.gardenClient.Client(), ms, gardencorev1beta1.GardenerName); err != nil {
-		return reconcile.Result{}, fmt.Errorf("could not ensure gardener finalizer: %w", err)
+	if !controllerutil.ContainsFinalizer(ms, gardencorev1beta1.GardenerName) {
+		r.getLogger(ms).Debug("Ensuring finalizer")
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_reconciler.go]: The original idea of "ensuring" instead of "adding" was: add the finalizer if it doesn't exist and do nothing otherwise. As such, the log message above should either be moved to the line before the check, or changed to `Adding finalizer`.

File: pkg/gardenlet/controller/bastion/bastion_reconciler.go [hunk 3]
```diff
@@ -131,7 +131,7 @@ func (r *reconciler) reconcileBastion(
 
 	// ensure finalizer is set
 	if !controllerutil.ContainsFinalizer(bastion, finalizerName) {
-		if err := controllerutils.PatchAddFinalizers(ctx, gardenClient, bastion, finalizerName); err != nil {
+		if err := controllerutils.StrategicMergePatchAddFinalizers(ctx, gardenClient, bastion, finalizerName); err != nil {
```
Comment (stoyanr) [inline pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: I think the following lines below are very dubious:
```
		// the patch above already triggers a reconcile, so we can stop here
		return nil
```
I think none of the other controllers has similar behavior. Might it be that it actually relies on the non-merge patch? Could you perhaps double-check this, or simply remove this `return nil` it looks somehow wrong anyway?

**Review by stoyanr — CHANGES_REQUESTED** [review]
Looks quite good, thanks! I have a couple of small comments and one general question: since checking for a finalizer and then adding exactly this finalizer is repeated so many times, wouldn't it be a better idea to do it via a helper function e.g. `EnsureFinalizer` in `finalizer.go`?

I would also like to test the change locally a bit before approving, e.g. by reconciling a seed and a shoot.

**Comment by stoyanr** [comment]
I tested creating / reconciling / deleting a shoot successfully with this change.

File: pkg/gardenlet/controller/bastion/bastion_reconciler.go (hunk 3 above)
Comment (timebertt) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: Nice catch, thank you!
TBH, I didn't pay much attention to this comment, but you're right. It seems like this code is on an update watch event caused by the finalizer.
I will remove the `return nil` to make the behavior similar to all other controllers.

Comment (timebertt) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: Ok, one step back. While addressing your general question, I noticed that the BackupEntry and Shoot controller are actually doing the same thing: https://github.com/gardener/gardener/blob/38ad556a7dec689b35c3ee6dd3cf59bd64b5c705/pkg/gardenlet/controller/shoot/shoot_control.go#L429-L434

And it seems to be for a good reason – to prevent unwanted double reconciliations, see https://github.com/gardener/gardener/pull/2364

Considering this, I think we should actually not merge `ContainsFinalizer` and `StrategicMergePatchAddFinalizers` to a single func, because we should actually distinguish between wether the finalizer was added or was already present.
And we should probably return after adding the finalizer (like it is done here and in the other 2 controlelrs) in all controllers.
WDYT?

Comment (timebertt) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: Alternatively, we could think about changing the "predicate" for enqueuing on update events:
instead of enqueuing when the generation is not equal to the observed generation, we could enqueue if the generation changed.
BTW: this is also what controller-runtime supports out of the box with `predicate.GenerationChangedPredicate`.

Comment (timebertt) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: Maybe @ialidzhikov has some additional thoughts on this as author of https://github.com/gardener/gardener/pull/2364

Comment (ialidzhikov) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: > And it seems to be for a good reason – to prevent unwanted double reconciliations, see #2364

Yes, as you stated the reason is to prevent any unwanted additional reconciliations. As described in https://github.com/gardener/gardener/pull/2364, on Shoot creation we have update/patch requests (like the one that adds a finalizer) that occur when metadata.generation != status.observedGeneration which is causing the obj to be requeued again. So currently on the Shoot reconciliation we use such "early exit" approach (`return nil` ) to prevent additional reconciliation (can be translated also like "exit the current reconciliation and rely on the next one that is caused by the updated that was just executed)".

You can check the description of https://github.com/gardener/gardener/pull/2364 for even better explanation.
So bottom line, we should make sure that we don't have (or introduce) additional reconciliations because of updates on existing creation/reconciliation.

PS. Since some time, in my local setup I notice again that on Shoot creation the Shoot gets reconciled again after a successful creation. I added this to my list, but still didn't manage to have a look.

Comment (stoyanr) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: It's a bit confusing for me that only a few controllers do this, and the others not. I would try to make it consistent across controllers, if it's not too hard. Does this even work properly with `StrategicMergePatchAddFinalizers` instead of `PatchAddFinalizers`?

> Alternatively, we could think about changing the "predicate" for enqueuing on update events:
instead of enqueuing when the generation is not equal to the observed generation, we could enqueue if the generation changed.

I like this idea, is it too complex a change? If yes, perhaps we could merge this PR as it is and do this refactoring in a separate PR.

Comment (timebertt) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: Thanks for your input!

> Does this even work properly with `StrategicMergePatchAddFinalizers` instead of `PatchAddFinalizers`?

I think, this should work as well with strategic merge patches. The only difference is, that we now don't get a conflict error anymore, if we read a stale object from the cache without a finalizer and then try to add it again. In that case we will simply return `nil` instead of a conflict error. So it will not go into exponential backoff but wait for the watch event instead before retrying. But this is also what happens currently with merge patch, when the first try to add the finalizer is successful and we return `nil` afterwards.

> So bottom line, we should make sure that we don't have (or introduce) additional reconciliations because of updates on existing creation/reconciliation.

> I like this idea, is it too complex a change? If yes, perhaps we could merge this PR as it is and do this refactoring in a separate PR.

I would like to do the following: we proceed with this PR as is (+ the commit for addressing the logging suggestions, will push it next), but we keep the current logic wrt to checking the finalizers first with `ContainsFinalizer` and then using `StrategicMergePatchAddFinalizers`. 
I will also keep the "special logic" in the shoot, bastion, etc. controller and not touch it there or in any other controllers to not regress anything in this PR.

And then, we can carefully think about the best way to prevent duplicated reconciliations in all controllers and check the duplicated shoot reconciliations that @ialidzhikov mentioned again.

Let me know, if you have any objections.

**Comment by timebertt** [comment]
I have pushed a new commit addressing the logging suggestions, but I'm keeping the current logic wrt to finalizer addition.
PTAL :)

File: pkg/gardenlet/controller/bastion/bastion_reconciler.go (hunk 3 above)
Comment (stoyanr) [reply pkg/gardenlet/controller/bastion/bastion_reconciler.go:134]: No objections, thanks!

**Review by stoyanr — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by timebertt** [comment]
Opened https://github.com/gardener/gardener/issues/4730 as a follow up
